phys2denoise/metrics/chest_belt.py

```diff
@@ -0,0 +1,217 @@
+
+    Parameters
+    ----------
+    belt_ts
+    window
+
+    Returns
+    -------
+    rpv_arr
+
+    Notes
+    -----
+    This metric was first introduced in [1]_.
+
+    1. Z-score respiratory belt signal
+    2. Calculate upper envelope
+    3. Calculate standard deviation of envelope
+
+    References
+    ----------
+    .. [1] J. D. Power et al., "Ridding fMRI data of motion-related influences:
+       Removal of signals with distinct spatial and physical bases in multiecho
+       data," Proceedings of the National Academy of Sciences, issue 9, vol.
+       115, pp. 2105-2114, 2018.
+    """
+    # First, z-score respiratory traces
+    resp_z = zscore(belt_ts)
+
+    # Collect upper envelope
+    rpv_upper_env = utils.rms_envelope_1d(resp_z, window)
+
+    # Calculate standard deviation
+    rpv_arr = np.std(rpv_upper_env)
+    return rpv_arr
+
+
+@due.dcite(references.POWER_2020)
+def env(belt_ts, samplerate, out_samplerate, window=10, lags=(0,)):
+    """Calculate respiratory pattern variability across a sliding window.
+
+    Parameters
+    ----------
+    belt_ts : (X,) :obj:`numpy.ndarray`
+        A 1D array with the respiratory belt time series.
+    samplerate : :obj:`float`
+        Sampling rate for belt_ts, in Hertz.
+    out_samplerate : :obj:`float`
+        Sampling rate for the output time series in seconds.
+        Corresponds to TR in fMRI data.
+    window : :obj:`int`, optional
+        Size of the sliding window, in the same units as out_samplerate.
+        Default is 6.
+    lags : (Y,) :obj:`tuple` of :obj:`int`, optional
+        List of lags to apply to the rv estimate. In the same units as
+        out_samplerate.
+
+    Returns
+    -------
+    env_arr
+
+    Notes
+    -----
+    This metric was first introduced in [1]_.
+
+    Across a sliding window, do the following:
+    1. Z-score respiratory belt signal
+    2. Calculate upper envelope
+    3. Calculate standard deviation of envelope
+
+    References
+    ----------
+    .. [1] J. D. Power et al., "Characteristics of respiratory measures in
+       young adults scanned at rest, including systematic changes and 'missed'
+       deep breaths," Neuroimage, vol. 204, 2020.
+    """
+    window = window * samplerate / out_samplerate
+    # Calculate RPV across a rolling window
+    env_arr = (
+        pd.Series(belt_ts).rolling(window=window, center=True).apply(rpv, window=window)
+    )
+    env_arr[np.isnan(env_arr)] = 0.0
+    return env_arr
+
+
+@due.dcite(references.CHANG_GLOVER_2009)
+def rv(belt_ts, samplerate, out_samplerate, window=6, lags=(0,)):
+    """Calculate respiratory variance.
+
+    Parameters
+    ----------
+    belt_ts : (X,) :obj:`numpy.ndarray`
+        A 1D array with the respiratory belt time series.
+    samplerate : :obj:`float`
+        Sampling rate for belt_ts, in Hertz.
+    out_samplerate : :obj:`float`
+        Sampling rate for the output time series in seconds.
+        Corresponds to TR in fMRI data.
```

Comment:
Do we want to include downsampling in the individual metric functions, or have it in a separate step?

Reply:
I'm thinking that downsampling should be done somewhere else TBH.